### PR TITLE
Feature: add sortable inventory

### PR DESCRIPTION
**Pull request by adrienlaws**
Adds

- User can move items with the mouse to sort them as desired
- Backend ability to re-order items in your inventory using MoveListElem from PR #484 
- Paves way for future inventory configuration options (maybe add preconfigured sort commands - group by reagents or alphabetical etc.)

Disclaimer / Notes

- Starting point was skittles1's PR below
- https://github.com/OpenMeridian/Meridian59/pull/964
- Thank to Hektic for helping me resolve some Git issues, and Zaphod for the optimized server side sort function.

Testing / QA

- I haven't seen any issues with this in limited testing
- Code looks sane
- Difficult to test alone in a comprehensive manner compared to production use

### CODE REVIEW

File: clientd3d/list.c [hunk 1]
```diff
@@ -237,6 +237,25 @@ void *list_find_item(list_type l, void *data, CompareProc compare)
    return NULL;
 }
 /************************************************************************/
+/*
+ * list_get_position: Find and return the position of an item in the given list.
+ *   The compare function is used to determine if two items are equal.
+ *   Returns NULL if item not found.
```
Comment (akirmse) [inline clientd3d/list.c:243]: Yikes, this is super broken.  NULL is the same as 0, so this function actually returns 0 for any item not in the list.

I suggest returning -1 instead, and checking for that value everywhere this is called.

File: clientd3d/list.c [hunk 2]
```diff
@@ -328,6 +347,85 @@ list_type list_move_to_front(list_type l, void *data, CompareProc compare)
    return l;
 }
 /************************************************************************/
+/*
+ * list_move_to_nth:  Find the given item in the given list, and
+ *   move it to the nth place.
+ *   Returns the list.
```
Comment (akirmse) [inline clientd3d/list.c:353]: This comment doesn't match what the function does.  For example, there's no argument "n".

File: kod/object/active/holder/nomoveon/battler/player/user.kod [hunk 3]
```diff
@@ -3749,6 +3758,51 @@ messages:
       return;
    }
 
+    UserMoveInventoryItem(what=$,where=$)
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/player/user.kod:3761]: "where" makes it sound like this is the ordinal position to move "what", but it's actually an *object* in the inventory where "what" will move.  Can you rename "where" (maybe to target_object or something better)?

File: kod/object/active/holder/nomoveon/battler/player/user.kod [hunk 4]
```diff
@@ -3749,6 +3758,51 @@ messages:
...
+      if what = $
+         OR where = $
+         OR what = where
+         OR plPassive = $
+      {
+         return;
+      }
+
+      % Get list position for the drop location.
+      iWhere = FindListElem(plPassive,where);
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/player/user.kod:3776]: There needs to be a check here in case "where" isn't found in plPassive.

File: kod/object/active/holder/nomoveon/battler/player/user.kod [hunk 5]
```diff
@@ -3749,6 +3758,51 @@ messages:
...
+      % MUST reassign this to plPassive, in case we're deleting the first
+      % list node. Use a temp list to verify there were no issues.
+      lSafeList = DelListElem(plPassive,what);
+      if lSafeList <> $
+      {
+         plPassive = lSafeList;
+      }
+      else
+      {
+         Debug("ALERT! Player ",self,Send(self,@GetTrueName)," tried to set ",
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/player/user.kod:3787]: This alert doesn't make sense to me.  When would it trigger?

File: kod/object/active/holder/nomoveon/battler/player/user.kod [hunk 6]
```diff
@@ -3749,6 +3758,51 @@ messages:
...
+
+      % MUST reassign this to plPassive, in case we're creating the first list node.
+      lSafeList = AddListElem(plPassive,iWhere,what);
+      if lSafeList <> $
+      {
+         plPassive = lSafeList;
+      }
+      else
+      {
+         Debug("ALERT! Player ",self,Send(self,@GetTrueName)," tried to set ",
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/player/user.kod:3799]: This also doesn't make sense to me.. how could this trigger?

File: module/merintr/inventry.c [hunk 7]
```diff
@@ -682,7 +683,14 @@ void InventoryLButtonUp(HWND hwnd, int x, int y, UINT keyFlags)
 
    // See if mouse pointer is in main graphics area
    if (!MouseToRoom(&temp_x, &temp_y))
+    {
+      // Drop in inventory. Check if we're still in inventory first.
+      if (x < 0 || y < 0)
```
Comment (akirmse) [inline module/merintr/inventry.c:688]: This looks a little fishy.  No check for maximum x or y values?  Does this only work because the inventory is in the lower right corner of the window?

**Review by akirmse — COMMENTED** [review]
This implementation is pretty weird.  It looks like it was originally written to move an inventory object to position N, but it actually moves an inventory object to the position of *another* inventory object.  I think I understand why it was done this way but the code as written isn't super clear.  Suggest clarifying this with some comments and renaming as described below.

File: clientd3d/list.c (hunk 1 above)
Comment (adrienlaws) [reply clientd3d/list.c:243]: It looks like there are a number of functions within this clientd3d/list.c file that return NULL

- list_first_item
- list_last_item
- list_delete_item
- list_delete
- list_destroy

Should I

1. make the requested change of returning -1 for the new list_get_position function only?
2. return -1 for list_get_position but also fix the other functions?
3. or not make this change so it returns NULL just as the other functions do?

Comment (akirmse) [reply clientd3d/list.c:243]: Other functions return NULL because it makes sense as a void * or list_type value.  They are not broken. 

However, this function returns an index, which is an int.  Since NULL is the same as 0, your #3 proposes returning 0, meaning, "item is in the first position in the list", when the item is not in the list.  That is surely not the behavior you want.  So I'd return -1 instead.

File: clientd3d/list.c (hunk 2 above)
Comment (adrienlaws) [reply clientd3d/list.c]: Renamed function to list_move_item and completely redid all the comments in a new commit.

File: clientd3d/list.c (hunk 1 above)
Comment (adrienlaws) [reply clientd3d/list.c]: Understood.  This is now fixed.  I have the function return -1 now, and references check for a -1 value.

File: kod/object/active/holder/nomoveon/battler/player/user.kod (hunk 5 above)
Comment (adrienlaws) [reply kod/object/active/holder/nomoveon/battler/player/user.kod]: Removed this section of code.  I'm speculating it was out of some sort of security concern.  It was part of the PR I imported.

File: kod/object/active/holder/nomoveon/battler/player/user.kod (hunk 6 above)
Comment (adrienlaws) [reply kod/object/active/holder/nomoveon/battler/player/user.kod]: Removed this section of code.  I'm speculating it was out of some sort of security concern.  It was part of the PR I imported.

File: clientd3d/list.c [hunk 8]
```diff
@@ -29,7 +29,7 @@ list_type list_create(void *newdata)
    return l;
 }
 /************************************************************************/
-/* list_add_item: add a node with data field newndata to end of l.  If l is NULL, 
+/* list_add_item: add a node with data field newdata to end of l.  If l is NULL,
```
Comment (adrienlaws) [inline clientd3d/list.c:32]: stumbled on this typo, figured I'd fix it

File: kod/object/active/holder/nomoveon/battler/player/user.kod (hunk 3 above)
Comment (adrienlaws) [reply kod/object/active/holder/nomoveon/battler/player/user.kod]: Changed parameters to oPayload and oTarget to match what I used in list.c on the client side.  Updated UserMoveInventoryItem function

File: kod/object/active/holder/nomoveon/battler/player/user.kod (hunk 4 above)
Comment (adrienlaws) [reply kod/object/active/holder/nomoveon/battler/player/user.kod]: Added checks for both payload and target objects, along with debug messages.

The debug messages fire and the function UserMoveInventoryItem returns without moving the items if either payload of target objects are not found in the plPassive list.

File: module/merintr/inventry.c (hunk 7 above)
Comment (adrienlaws) [reply module/merintr/inventry.c]: Added additional checks including one that verifies that when the user lets go of the left mouse button, they are within the inventory window area per InventoryGetArea() and IsInArea() functions

File: blakcomp/function.c [hunk 9]
```diff
@@ -36,6 +36,7 @@ function_type Functions[] = {
 {"Nth",			NTH,		 AEXPRESSION,	AEXPRESSION,	ANONE},
 {"SetFirst",            SETFIRST,        AEXPRESSION,   AEXPRESSION,    ANONE},
 {"SetNth",              SETNTH,          AEXPRESSION,   AEXPRESSION,    AEXPRESSION, ANONE},
+{"SwapListElem", SWAPLISTELEM, AEXPRESSION, AEXPRESSION, AEXPRESSION, ANONE},
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply blakcomp/function.c:39]: I think I have messed this up and I need to adjust the server side function to be insert instead of swap. More info below.

On the client side list.c, the item the user drags (item1 or payload) is inserted at the position of item2 (target). So for example if item1 (payload) is in position 3 and item2 (target) is in position 7, then item1 will be removed from the list, item2 will drop to position 6, and then item1 will be inserted at item2's original position of 7. This is all what we're looking for.

The server receives the request to move the items, so the client inventory and the server inventory will match, but on the server side I am using a swap function. This is problematic.

In short, I need to take a look at it. And it probably makes sense to make the server side function an insert function, because otherwise the algorithms aren't the same even if they're targeting the same items. This will probably result in unpredictable behavior when the user logs out and in, and sees their inventory reordered to something they didn't expect. I have tested it and sometimes that happens but it's unclear why (lag, packet not received, or the sort functions not matching).

[the remark this answers is not in the document]
Comment (adrienlaws) [reply blakcomp/function.c]: Implemented new sort function MoveListElem from other PR #484

File: blakserv/list.c [hunk 10]
```diff
@@ -224,6 +224,52 @@ int SetNth(int n,int list_id,val_type new_val)
 	return NIL;
 }
 
+int SwapListElem(int list_id,int elem_one,int elem_two)
+{
+   list_node *l, *list_node_one, *list_node_two;
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply blakserv/list.c]: The SwapListElem function has been removed and replaced by the new MoveListElem function.  Resolving related conversation.

File: kod/object/active/holder/nomoveon/battler/player/user.kod [hunk 11]
```diff
@@ -3749,6 +3758,48 @@ messages:
...
+      if iPayload = $
+         {
+            Debug("ALERT! ",Send(self,@GetTrueName),self,"tried to move "
+                  "an object in their inventory that does not exist!");
+
+            return;
+         }
+
+      % Check that the target object isn't found in plPassive.
+      if iTarget = $
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply kod/object/active/holder/nomoveon/battler/player/user.kod]: updated and combined debug messages for UserMoveInventoryItem as requested

File: module/merintr/inventry.c [hunk 12]
```diff
@@ -671,18 +672,32 @@ void InventoryLButtonDown(HWND hwnd, BOOL fDoubleClick, int x, int y, UINT keyFl
...
+ */    
 void InventoryLButtonUp(HWND hwnd, int x, int y, UINT keyFlags)
 {
    int temp_x, temp_y;
    room_contents_node *r;
+   POINT mouse;
+   AREA a;
 
-   if (!InventoryReleaseCapture())
+   // Multiple Sanity Checks
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply module/merintr/inventry.c]: Removed several comments as requested, resolving relevant conversations

File: module/merintr/inventry.c [hunk 13]
```diff
@@ -671,18 +672,32 @@ void InventoryLButtonDown(HWND hwnd, BOOL fDoubleClick, int x, int y, UINT keyFl
...
+   if (!InventoryReleaseCapture()) // Check that we're still in inventory
       return;
 
-   // See if mouse pointer is in main graphics area
-   if (!MouseToRoom(&temp_x, &temp_y))
+   InventoryGetArea(&a);                  // Get inventory area
+   GetCursorPos(&mouse);                  // Get mouse position
+   ScreenToClient(cinfo->hMain, &mouse);  // Get mouse position in client coordinates
+      
+   if (!MouseToRoom(&temp_x, &temp_y))       // If the mouse pointer is NOT in the main graphics area
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply module/merintr/inventry.c]: Streamlined checks.  Removed extra checks and consolidated around IsInArea

File: module/merintr/inventry.c [hunk 14]
```diff
@@ -671,18 +672,32 @@ void InventoryLButtonDown(HWND hwnd, BOOL fDoubleClick, int x, int y, UINT keyFl
...
+    {
+      if (x < 0 || y < 0)                    // Another check that we are in inventory
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply module/merintr/inventry.c]: Removed this check.  Consolidated around IsInArea check

File: module/merintr/inventry.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (adrienlaws) [reply module/merintr/inventry.c]: This one is old, and is needed.

The function InventoryLButtonUp first checks if we are in the inventory area, and if so moves items around and exits the function.

Then this old check sees if we are not in the main window area (we already established we are not in the inventory area).  It will exit if so.  The reason is that the next few lines of code check for dropping items and putting items into a container.  It should only do so if we are not in the inventory window AND we are in the main 3D window.  If we are not, then the function exits.

TLDR; this is old and needed.

File: module/merintr/inventry.c [hunk 15]
```diff
@@ -1062,6 +1077,38 @@ Bool InventoryDropCurrentItem(room_contents_node *container)
    return True;
 }
 /************************************************************************/
+/*
+ * InventoryMoveCurrentItem:  Move the item with the inventory cursor, if any.
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply module/merintr/inventry.c:1080]: Updated the documentation as requested to state (0,0) is the top left corner of the inventory window.  To the right -> increases x, while down increases y.  This is standard along with all the other functions in inventry.c

File: module/merintr/inventry.c [hunk 16]
```diff
@@ -1062,6 +1077,38 @@ Bool InventoryDropCurrentItem(room_contents_node *container)
...
+   row = top_row + y / INVENTORY_BOX_HEIGHT;
+
+   drop_position = (InvItem *) list_nth_item(items, row * cols + col);
+   if (drop_position == NULL)
+      return False;
+
+   if (item->obj->id == drop_position->obj->id)
+      return False;
+
+   items = list_move_item(items,(void *)item->obj->id,(void *)drop_position->obj->id,InventoryCompareIdItem);
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply module/merintr/inventry.c]: Added spaces after commas as requested.

File: module/merintr/inventry.c [hunk 17]
```diff
@@ -1062,6 +1077,38 @@ Bool InventoryDropCurrentItem(room_contents_node *container)
    return True;
 }
 /************************************************************************/
+/*
+ * InventoryMoveCurrentItem:  Move the item with the inventory cursor, if any.
+ *   Return True iff item moved.
+ */
+Bool InventoryMoveCurrentItem(int x, int y)
+{
+   int row, col;
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply module/merintr/inventry.c]: Moved declaration further down the function to where they are used, as requested.

File: clientd3d/list.c [hunk 18]
```diff
@@ -328,6 +347,97 @@ list_type list_move_to_front(list_type l, void *data, CompareProc compare)
    return l;
 }
 /************************************************************************/
+/*
+ * list_move_item:  Given a list and two items, removes payload node (item1) and inserts
+ *   it at target index (item2's original index position). Returns a new list if the items
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply clientd3d/list.c]: Updated all references to item1 and item2 to data1 and data2.  Removed all references to item1 & 2

File: clientd3d/list.c [hunk 19]
```diff
@@ -328,6 +347,97 @@ list_type list_move_to_front(list_type l, void *data, CompareProc compare)
...
 }
 /************************************************************************/
+/*
+ * list_move_item:  Given a list and two items, removes payload node (item1) and inserts
+ *   it at target index (item2's original index position). Returns a new list if the items  
+ *   are not in the same position. Otherwise returns the original list.
+ */
+list_type list_move_item(list_type l, void *data1, void *data2, CompareProc compare)
+{
+   list_type payload_data, prev, temp;
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply clientd3d/list.c]: Moved declarations to right before they are used

File: clientd3d/list.c [hunk 20]
```diff
@@ -328,6 +347,97 @@ list_type list_move_to_front(list_type l, void *data, CompareProc compare)
...
+
+   int i;
+
+   /* Sanity checks */
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply clientd3d/list.c]: Removed as requested.

File: clientd3d/list.c [hunk 21]
```diff
@@ -328,6 +347,97 @@ list_type list_move_to_front(list_type l, void *data, CompareProc compare)
...
+   if (l == NULL)
+      return NULL;
+
+   int pos_payload = list_get_position(l, data1, compare);
+   int pos_target  = list_get_position(l, data2, compare);
+
+   /* Check that payload (item1) and target (item2) are actually in the list */
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply clientd3d/list.c]: Changed all single line instances to C++ style comments //

File: clientd3d/list.c [hunk 22]
```diff
@@ -328,6 +347,97 @@ list_type list_move_to_front(list_type l, void *data, CompareProc compare)
...
+   if (pos_payload == -1)
+      return l;
+
+   if (pos_target == -1)
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply clientd3d/list.c]: Done, combined.

File: clientd3d/list.c [hunk 23]
```diff
@@ -328,6 +347,97 @@ list_type list_move_to_front(list_type l, void *data, CompareProc compare)
...
+      return l;
+
+   /* If payload and target position are the same, return the original list */
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply clientd3d/list.c]: Removed as requested.

File: clientd3d/list.c [hunk 24]
```diff
@@ -328,6 +347,97 @@ list_type list_move_to_front(list_type l, void *data, CompareProc compare)
...
+   if (pos_payload == pos_target)
+      return l;
+
+   /* Check if target (item2) is at the beginning of the list.  If so, use existing function list_move_to_front() */
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply clientd3d/list.c]: Removed as requested.

File: clientd3d/list.c [hunk 25]
```diff
@@ -328,6 +347,97 @@ list_type list_move_to_front(list_type l, void *data, CompareProc compare)
...
+      }
+      prev = temp;
+      temp = temp->next;
+   }
+
+   // If the payload (item1) was never found in the list, return the original list
+   if (payload_data == NULL)   
+      return l; 
+
+   /* We now have the value of the payload (item1) in payload_data, we have removed the payload node from the list l, and we know the target (item2) original index position pos_target.  We now need to re-insert the payload (item1) at the target index (item2's original index position).  We do this by walking the list and re-inserting it at the target index of item2.*/
```
[the remark this answers is not in the document]
Comment (adrienlaws) [reply clientd3d/list.c]: Wrapped comment as requested.
